File: RubyWrapperCielo24/lib/RubyWrapperCielo24/web_utils.rb
module RubyWrapperCielo24
  class WebUtils

    require 'json'
    require 'httpclient'

    VERIFY_MODE = nil
    BASIC_TIMEOUT = 60
    DOWNLOAD_TIMEOUT = 300

    def self.get_json(uri, method, timeout, query=nil, headers=nil, body=nil)
      response = http_request(uri, method, timeout, query, headers, body)
      return JSON.parse(response)
    end

    def self.http_request(uri, method, timeout, query=nil, headers = nil, body = nil)
      http_client = HTTPClient.new
      http_client.cookie_manager = nil
      # TODO: timeout

      response = http_client.request(method, uri, query, body, headers, nil)

      # Handle web errors
      if response.status_code == 200 or response.status_code == 204
        return response.body
      else
        json = JSON.parse(response.body)
        raise WebError.new(json["ErrorType"], json["ErrorComment"])
      end
    end
  end

  class WebError < StandardError
    attr_reader :type
    def initialize(type, comment)
      super comment
      @type = type
    end

    def to_s
      return @type + " - " + super.to_s
    end
  end
end

File: cielo24_PHP/src/cielo24/WebUtils.php
<?php

class WebUtils {
    const BASIC_TIMEOUT = 60;           # seconds
    const DOWNLOAD_TIMEOUT = 300;       # seconds
    const UPLOAD_TIMEOUT = 604800;      # seconds (7 days)

    public static function getJson($base_uri, $path, $method, $timeout, $query=array(), $headers=array(), $body=null) {
        $response = WebUtils::httpRequest($base_uri, $path, $method, $timeout, $query, $headers, $body);
        return json_decode($response.getBody());
    }

    public static function httpRequest($base_uri, $path, $method, $timeout, $query=array(), $headers=array(), $body=null) {
        if ($query == null){
            $query = array();
        }
        if ($headers == null) {
            $headers = array();
        }

        $url = $base_uri . $path;
        if (count($query) > 0) {
            $url .= "?" . http_build_query($query);
        }
        $http_request = new HttpRequest($url, $method, array("timeout" => $timeout));
        $http_request.setHeaders($headers);
        $http_request.setBody($body);
        $response = $http_request.send();

        if ($response.getResponseCode() == 200 or $response.getResponseCode() == 204) {
            return $response.getBody();
        } else {
            $json = json_decode($response.getBody());
            throw new WebError($json["ErrorType"], $json["ErrorComment"]);
        }
    }
}

class WebError extends Exception {

    public $errorType;
    public $errorComment;

    public function __construct1($type, $comment) {
        $this->errorType = $type;
        $this->errorComment = $comment;
    }

    public function __toString() {
        return $this->errorType . " - " . $this->errorComment;
    }
}